---
title: 'Microarray Data Analysis -- GSE19187 uncontrolled_asthma vs healthy'
author: 'Mengyuan Kan (mengykan@upenn.edu)'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: FALSE
    depth: 3
---
***

This report is based on the differential analysis of microarray data. The main steps include:

* Get raw data from a GEO study
* Normalize gene expression data
* Perform gene differential expression analysis
* Adjust for batch effect
* Visualize gene expression results

The requisitions for input:

* A phenotype file derived from the GEO data

Outputs:

* Gene differential expression analysis result file(s) (.csv) by tissue
* A gene expression analysis report (.html)

Change the variables here
```{r, eval=T}
# GEO id
geo_id="GSE19187"
# phenotype file
pheno_fn="C:/Users/mengykan/Projects/AsthmaApp/res/microarray/GSE19187_NE_uncontrolled_asthma_vs_healthy.txt"
# direcotry stores GEO data
datadir="C:/Users/mengykan/Projects/AsthmaApp/data/microarray"
# directory stores differential gene expression analysis results and report file
resdir="C:/Users/mengykan/Projects/AsthmaApp/res/microarray"
# reference condition
con0="healthy"
# altered condition
con1="uncontrolled_asthma"
```

Install the prerequisite R packages
```{r eval=F}
source("http://bioconductor.org/biocLite.R")
biocLite("GEOquery")
biocLite("oligo")
biocLite("limma")
biocLite("hgu133plus2.db")
biocLite("hugene10sttranscriptcluster.db")
biocLite("sva")
biocLite("annotate")
biocLite("viridis")
install.packages("gplots")
install.packages("ggplot2")
install.packages("devtools")
install.packages("pander")
```

Load the necessary libraries
```{r eval=T, message=F, warning=F}
library(GEOquery)
library(oligo)
library(limma)
library(sva)
library(annotate)
library(viridis)
library(gplots)
library(ggplot2)
library(devtools)
library(pander)
```

### Obtain Gene Expression Raw Data and Assign Phenotype Inforamtion

Download GEO files if the data folder does not exist.
```{r eval=T,  message=FALSE}
if (!(dir.exists(paste0(datadir,"/",geo_id)))) {
  getGEOSuppFiles(geo_id,baseDir=datadir) #download GEO files
  untar(paste0(datadir,"/",geo_id,"/",geo_id,"_RAW.tar"), exdir=paste0(datadir,"/",geo_id,"/data")) # extract the zip file
}
```

Read in raw data
```{r eval=T,  message=FALSE, warning=F, results="hide"}
celFiles <- list.celfiles(paste0(datadir,"/",geo_id,"/data/"), full.names=TRUE, listGzipped=TRUE)
raw.data <- read.celfiles(celFiles)
```

Read in pre-prepared phenotype data
```{r eval=T, message=FALSE, results="asis"}
pheno <- read.table(pheno_fn, header=TRUE, sep="\t") # read in the phenotype data
dim(pData(raw.data))
dim(pheno)
```

Subset raw data by selecting samples in phenotype file and select samples that passed QC
```{r eval=T, results="asis"}
raw.data.sub <- raw.data[, sampleNames(raw.data)%in%pheno$Filename] # subset raw data with samples in phenotype file
pData(raw.data.sub) <- pheno # assign phenotype information to raw data
row.names(pData(raw.data.sub)) <- sampleNames(protocolData(raw.data.sub)) # assign row names
raw.data.of.interest <- raw.data.sub[,raw.data.sub$QC_Pass==1] # select samples that passed QC
# show phenotype subset
res <- pData(raw.data.of.interest)
row.names(res) <- NULL
pandoc.table(res,split.table = Inf)
```

Determine the comparison type (disease or treatment status) and assign the comparison column to a new column "Status"
```{r eval=T}
for (comptype in c("Disease","Treatment")[c("Disease","Treatment")%in%colnames(pData(raw.data.of.interest))]) { # check if Disease and/or Treatment columns are in the phenotype file
  if (con0%in%pData(raw.data.of.interest)[,comptype]&con1%in%pData(raw.data.of.interest)[,comptype]){ # check if both conditions are in Disease/Treatment column and determine the comparison type
    raw.data.of.interest$Status <- pData(raw.data.of.interest)[,comptype] # create a new column named as Status
    break
  }
}
# set the order levels for comparison status (1st: con1, 2nd: con0)
raw.data.of.interest$Status <-factor(raw.data.of.interest$Status,levels=c(con0,con1))
```

Check the sample size in each batch and status
```{r eval=T, results="asis"}
res <- as.data.frame(table(pData(raw.data.of.interest)[,c("Status")]))
names(res) <- c("Status","Frequency")
pandoc.table(res,split.table=Inf)
res <- as.data.frame(table(pData(raw.data.of.interest)[,c("Status","ScanDate_Group")]))
names(res) <- c("Status","ScanDate_Group","Frequency")
pandoc.table(res,split.table=Inf)
```

Assign colours to status and scan date
```{r eval=T}
# assign colours to comparison status
colour_status <- c("navy","red")
names(colour_status) <- c(con0,con1) # assign red to condition 1 and navy to condition 2
colour_status_list <- unlist(lapply(raw.data.of.interest$Status,function(x){colour_status[x]}))
# assign colours to scan date
colours=c("#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666", "#8DD3C7", "#FFFFB3", "#BEBADA", "#FB8072", "#80B1D3", "#FDB462", "#B3DE69", "#FCCDE5", "#D9D9D9", "#BC80BD", "#CCEBC5", "#FFED6F") # first 8 colour names derived from Dark2, and last 12 names from Set3
raw.data.of.interest$ScanDate_Group <- as.factor(raw.data.of.interest$ScanDate_Group)
i=nlevels(raw.data.of.interest$ScanDate_Group)
colour_scandate <- colours[1:i]
names(colour_scandate) <- levels(raw.data.of.interest$ScanDate_Group) # colour to corresponding scan date
colour_scandate_list <- unlist(lapply(raw.data.of.interest$ScanDate_Group,function(x){colour_scandate[x]}))
```

Create a summary table for this comparison
```{r eval=T, results="asis"}
name <- gsub("\\..*","",basename(pheno_fn)) # directly adopt the phenotype file name as a unique id for this comparison study
N_Condition0 <- sum(raw.data.of.interest$Status==con0) # count number of samples under condition 0
N_Condition1 <- sum(raw.data.of.interest$Status==con1) # count number of samples under condition 1
if (paste0(comptype,"_des")%in%names(pData(raw.data.of.interest))) { # if description column exsists in the phenotype file
  if (comptype=="Disease") {Description=levels(as.factor(pData(raw.data.of.interest)$Disease_des))}
  if (comptype=="Treatment") {Description=levels(as.factor(pData(raw.data.of.interest)$Treatment_des))}
} else {Description=con1}
df <- data.frame(
  GEO_ID=geo_id,
  Tissue=levels(raw.data.of.interest$Tissue),
  App=NA,
  Disease=NA,
  Treatment=NA,
  Condition0=con0,
  Condition1=con1,
  N_Condition0=N_Condition0,
  N_Condition1=N_Condition1,
  Total=N_Condition0+N_Condition1,
  Description=Description,
  Unique_ID=name
)
# create App variable based on the type of comparison
# if a comparison is based on the disease status, treatment variable is the combination of comparison status, otherwise use the variable in the phenotype file.
if (comptype=="Disease") {df$App="Disease";df$Disease <- paste0(con1,"_vs_",con0)} else if ("Disease"%in%colnames(phenoData(raw.data.of.interest))) {df$Disease <- unique(phenoData(raw.data.of.interest)$Disease)}
# if a comparison is based on the disease status, treatment variable is the combination of comparison status, otherwise use the variable in the phenotype file.
if (comptype=="Treatment") {df$App="Treatment";df$Treatment <- paste0(con1,"_vs_",con0)} else if ("Treatment"%in%colnames(phenoData(raw.data.of.interest))) {df$Treatment <- unique(phenoData(raw.data.of.interest)$Treatment)}
pandoc.table(t(df),split.table=Inf)
```

Save the summary table in .csv format
```{r eval=T}
write.csv(df, file = paste0(resdir,"/",name,"_infosheet.csv"), row.names = FALSE)
```

### Differential Gene Expression Analysis
#### Normalize raw gene expression data
Normalize gene expression raw data using robust multi-array average (RMA) method
```{r eval=T,message=FALSE, results="hide"}
rma.data.of.interest = rma(raw.data.of.interest) # perform the RMA
cols=colour_status_list
boxplot(raw.data.of.interest,target="core",col=cols,main="Raw Probe Intensities",xaxt="n") # view raw data
legend("topright",legend=names(colour_status),fill=colour_status)
boxplot(rma.data.of.interest,col=cols,main="RMA Probe Intensities",xaxt="n") # view RMA-adjusted data
legend("topright",legend=names(colour_status),fill=colour_status)
```


#### Pair-wise comparison between status
Create a design model matrix for linear model. Fit a linear model using design matrix model.
```{r eval=T, results="asis"}
design = model.matrix(~ -1+factor(rma.data.of.interest$Status)) # create design model matrix
colnames(design) = levels(factor(rma.data.of.interest$Status))
pandoc.table(data.frame(Sample=rma.data.of.interest$Sample,design),split.table = Inf)
fit = lmFit(rma.data.of.interest, design) # fit a linear model to estimate the fold changes and standard error
```

Create a contrast group and fit it in a linear model
```{r eval=T,results="asis"}
data.contrast = makeContrasts(contrasts=paste(c(con1,con0),collapse="-"),levels = design) # create a contrast group by comparing con1 vs con0
pandoc.table(as.data.frame(data.contrast))
fit2 = contrasts.fit(fit, data.contrast) # get the contrasts for samples of interest
fit2 = eBayes(fit2) # adjust fit coefficients using an empirical Bayes moderation of standard errors
contrast_table <- topTable(fit2, adjust="BH",num=Inf) # get full set of results for each hypothesis test can be extracted using
```

#### Adjust p-values with batch effect

Check if there are multiple batches in the study.
```{r eval=T, results="asis"}
nbatch=nlevels(rma.data.of.interest$ScanDate_Group)
```

Create a full model 
```{r eval=T, results="asis"}
modBatch =""
if (nbatch>1) {
 modBatch = model.matrix(~factor(rma.data.of.interest$Status)+factor(rma.data.of.interest$ScanDate_Group) ) # full model (adjustment variables and variables of interest)
 res <- as.data.frame(modBatch)
 names(res) <- c("Intercept",levels(rma.data.of.interest$Status)[-1],levels(rma.data.of.interest$ScanDate_Group)[-1])
 pandoc.table(res,split.table=Inf)
}
```

Create a null model only including batch variables
```{r eval=T, results="asis"}
nullBatch=""
if (nbatch>1) {
  nullBatch =  model.matrix(~factor(rma.data.of.interest$ScanDate_Group)) # null model (adjustment variables only)
  res <- as.data.frame(nullBatch)
  names(res) <- c("Intercept",levels(rma.data.of.interest$ScanDate_Group)[-1])
  pandoc.table(res,split.table=Inf)
}
```

Check if batch and comparison status are confounded

As SVA computes matrix x in t(batch model)%*%x=batch model, batch effect adjustment can be performed only when the solve function works and assign TRUE to the object "batchadj", otherwise, the batch and status are highly confounded and assign FLASE to "batchadj.
```{r eval=T}
batchadj=""
if (nbatch>1) {
 if (class(try(solve(t(modBatch)%*%modBatch),silent=T))=="matrix") {batchadj=TRUE} else {batchadj=FALSE; message("The batch and comparison status are highly confounded")}
}
```

Compute f statistic p-values using SVA and q-values if batch and status are not confounded.

If batch and status are confounded, assign NA to the batch adjusted p-values.

If there is only one batch, assign original p-values to the batch adjusted p-values

```{r eval=T, results="asis"}
if (nbatch>1) {
  if (batchadj) {
    pValuesBatch = f.pvalue(exprs(rma.data.of.interest), modBatch, nullBatch) # get batch effect-adjusted p-values
    qValuesBatch = p.adjust(pValuesBatch, method="BH") # get q-values
    contrast_table_adjust <- merge(contrast_table,data.frame(pValuesBatch,qValuesBatch),by="row.names", all = TRUE) # merge adjusted p and q-values in previous results
  } else {
    Row.names <- row.names(contrast_table)
    pValuesBatch <- rep(NA,nrow(contrast_table))
    qValuesBatch <- rep(NA,nrow(contrast_table))
    contrast_table_adjust <- data.frame(Row.names=Row.names,contrast_table,pValuesBatch=pValuesBatch,qValuesBatch=qValuesBatch)
  } 
} else {
  Row.names <- row.names(contrast_table)
  pValuesBatch <- contrast_table$P.Value
  qValuesBatch <- contrast_table$adj.P.Val
  contrast_table_adjust <- data.frame(Row.names=Row.names,contrast_table,pValuesBatch=pValuesBatch,qValuesBatch=qValuesBatch)
}
pandoc.table(summary(contrast_table_adjust),caption=paste0(con1," vs. ",con0," summary"),split.table = Inf)
```


#### Assign official gene symbol

Assign official gene symbol using hgu133plus2.db package if anntoation is based onpd.hg.u133a.

Assign official gene symbol using hgu133plus2.db package if anntoation is based on pd.hg.u133.plus.2. Note that the probe names have additional "PM"s in the results will be removed to match the names in the gene symbol package.

Assign official gene symbol using hugene10sttranscriptcluster.db if annotation is based on pd.hugene.1.0.st.v1.

```{r eval=T, warning=F,message=F, results="asis"}
if (grepl("u133",annotation(raw.data))) {
  library(hgu133plus2.db)
  remove.pm <- function(x) {gsub("_PM", "", x)}
  contrast_table_adjust$ID = sapply(contrast_table_adjust[,1], remove.pm)
  contrast_table_adjust$SYMBOL <-
lapply(as.character(contrast_table_adjust$ID), function(x) mget(x,
env=hgu133plus2SYMBOL, ifnotfound=NA)[[1]])} else if (grepl("hugene",annotation(raw.data))) {
  library(hugene10sttranscriptcluster.db)
  contrast_table_adjust$ID = contrast_table_adjust[,1]
  contrast_table_adjust$SYMBOL <- lapply(as.character(contrast_table_adjust$ID), function(x) mget(x, env =hugene10sttranscriptclusterSYMBOL, ifnotfound=NA)[[1]])
}
contrast_table_adjust <- contrast_table_adjust[order(contrast_table_adjust$P.Value),]
```

Save the differential gene expression analysis results
```{r eval=T, message=FALSE}
res <- contrast_table_adjust
res <- data.frame(lapply(res, as.character)) # make sure dataframes do not contain list. Take out "striresssngsAsFactors=FALSE" from previous script
name=gsub("\\..*","",basename(pheno_fn)) # directly adopt the phenotype file name for result file name
write.csv(res, file = paste0(resdir,"/",name,".csv"), row.names = FALSE)
```

### Visualize Gene Expression Results

#### View top 20 differentially gene expression results

Top 20 q-values not adjusted for batch effect
```{r eval=T,results="asis"}
res=contrast_table_adjust[1:20,]
dat_reform <- function(dt) {
  round2 <- function(x){round(x,2)}
  dt[,c("logFC","AveExpr","t","B")] <- sapply(dt[,c("logFC","AveExpr","t","B")],round2)
  sciform <- function(x){format(x,,scientific=TRUE,digits =2)}
  dt[,c("P.Value","adj.P.Val","pValuesBatch","qValuesBatch")] <- sapply(dt[,c("P.Value","adj.P.Val","pValuesBatch","qValuesBatch")],sciform)
  dt
}
res <- dat_reform(res)
rownames(res) <- NULL
pandoc.table(res,split.table=Inf)
```

Top 5 q-values with batch effect adjusted if batch effect was adjusted.
```{r eval=T,results="asis"}
if (all(is.na(contrast_table_adjust$pValuesBatch))) {
  message("The batch and status are not confounded. Batch effect was not adjusted.")} else if (nbatch==1) {
  message("There is only one batch in this study.")} else {
  res <- contrast_table_adjust[order(contrast_table_adjust$pValuesBatch),][1:5,]
  res <- dat_reform(res)
  rownames(res) <- NULL
  pandoc.table(res,split.table=Inf)
}
```

#### Boxplots for top 5 differentially expressed genes

This step helps to visualize and check the effect direction in the comparison
```{r eval=T}
res=contrast_table_adjust[1:5,]
for (j in 1:5) {
  probe_top <- res[j,1]
  gene_top <- res$ID[j]
  values=exprs(rma.data.of.interest)[probe_top, ] 
  status=rma.data.of.interest$Status
  df <- data.frame(values=values,status=status)
  title=paste0("top ",j," probe ", gene_top," expression")
  print(
      ggplot(df,aes(x=status,y=values)) +
        geom_boxplot(outlier.colour=NA,color="grey18",fill=colours[1]) +
        stat_boxplot(geom ='errorbar', color="grey18") +
        geom_jitter(size=1,position = position_jitter(width=0.3)) +
        labs(title=title) +
        theme_bw() +
        theme(legend.position="none",axis.title=element_blank())
  )
}
```

#### Volcano plot

Create a volcano plot function
```{r eval=T}
volplot <- function(df,qval_column) {
  # get qvalue column
  qval <- df[,qval_column]
  if (all(is.na(qval))) {message("The batch and status are not confounded. Batch effect is not adjusted.")} else {
    if (min(qval)>=0.05) {
      message(paste0("No differentially expressed genes detected with a ",qval_column," <0.05"))
    } else {
      # remove NA values
      df <- df[!is.na(qval),]
      # assign colors to DE and non-DE genes
      df$sig <- rep(NA,nrow(df))
      df$sig[qval<0.05] <- "red"
      df$sig[qval>=0.05] <- "black"
      df$sig <- as.factor(df$sig)
      color <- levels(df$sig)
      # log10 transformed q values
      df$logqval <- -log10(qval)
      ggplot(df, aes(x = logFC, y = logqval, color=sig)) + geom_point() +
        theme_bw() +
        labs(x="logFC",y="-log10(q-value)") +
        scale_color_manual(values=color) +
        theme(legend.position="none")
    }
  }
}
```

Volcano plot if there are genes with a qvalue <0.05

```{r eval=T}
volplot(contrast_table_adjust,"adj.P.Val")
```

Volcano plot if there are genes with a batch adjusted qvalue <0.05

```{r eval=T}
if (nbatch==1) {message("There is only one batch in this study.")} else {
  volplot(contrast_table_adjust,"qValuesBatch")
}
```

#### Heatmap for top 200 differentially expressed gene probes

Show colours by comparison status

```{r, eval=T, fig.height=10, fig.width=12}
top.rma <- rma.data.of.interest[row.names(rma.data.of.interest)%in%contrast_table_adjust[1:200,1]] # plot heatmap for top 200 genes
status.cols=colour_status_list # assign colors to status
heatmap_topgene <- function(dt,status.cols) {
  array_name <- gsub(".CEL.gz","",colnames(dt)) # shorten the sample id
  heatmap.2(exprs(top.rma), col=viridis(256, option="B"),
ColSideColors=status.cols,
    labCol=array_name,labRow = "", # take out gene probe id
    trace="none",
    margins=c(12,20), # (bottom margin, left margin)
    cexRow=1,cexCol=1,
    keysize=1.5,key.title=NA,key.xlab="Gene Expression Values",key.ylab="Counts")
}
heatmap_topgene(top.rma,colour_status_list) # assign colors to status
legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.6)
```

Show colours by scan date

```{r, eval=T, fig.height=10, fig.width=12}
heatmap_topgene(top.rma,colour_scandate_list) # assign colors to scan date
legend("bottomleft",legend=names(colour_scandate),fill=colour_scandate,cex=0.6)
```

#### Heatmap for sample correlation

Compute sample correlation in a matrix
```{r, eval=T}
# compute correlation matrix
rma.data.cor <- cor((exprs(rma.data.of.interest)))
```

Show colours by comparison status
```{r, eval=T, fig.height=10, fig.width=12}
cor_plot <- function(m,colorlist) {  # m: correlation matrix, colorlist: assign colours to row and column
  # compute distance based on 1-correlation efficient
  dend = as.dendrogram(hclust(as.dist(1-m), method = "single"))
  ord = order.dendrogram(dend)
  # rename samples
  array_name <- colnames(m)
  array_name <- gsub(".CEL.gz","",array_name) # shorten the sample id
  # heatmap plot
  heatmap.2(m,,Rowv=dend,Colv=dend,
    col=viridis(256,option="B"),ColSideColors=colorlist,RowSideColors=colorlist,
    labCol=array_name,labRow=array_name,
    trace="none",
    margins=c(12,20), # (bottom margin, left margin)
    cexRow = 1,cexCol = 1,
    keysize=1.5,key.title=NA,key.xlab="Dist2",key.ylab="Counts")
}
cor_plot(rma.data.cor,colour_status_list)
legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.6)
```

Show colours by scan date
```{r, eval=T, fig.height=10, fig.width=12}
cor_plot(rma.data.cor,colour_scandate_list)
legend("bottomleft",legend=names(colour_scandate),fill=colour_scandate,cex=0.6)
```

#### Principal component analysis (PCA)

Compute PCs and variance explained by the first 10 PCs
```{r eval=T, results="asis"}
# obtain original expression data
rma.data.pca <- na.omit(exprs(rma.data.of.interest))
# As scale function divides by the variance, the probe with the expression sd=0 across samples must be removed.
sd <- apply(rma.data.pca,1,sd)
rma.data.pca <- rma.data.pca[!sd==0,]
# compute pcs
pca <- prcomp(t(rma.data.pca), retx = TRUE, center = TRUE, scale = TRUE)
pc <- data.frame(pca$x)
# compute variance explained by each PC
vars <- pca$sdev^2
pcs <- t(pc)
pvars <- vars*100.0/sum(vars) # proportion of variance (%) explained by each PC
cumsum_pvars <- cumsum(pvars) # Cumulative Proportion of Variance (%)
if (nrow(pcs)>10) {nres <- 10} else {nres=nrow(pcs)} # select top 10 PCs if number of PCs >10
res <- data.frame(rownames(pcs),pvars,cumsum_pvars)[1:nres,]
names(res) <- c("PC","Proportion of Variance (%)","Cumulative Proportion of Variance (%)")
pandoc.table(res,split.table = Inf)
```

Plot first two PCs coloured by status.
```{r eval=T}
df <- data.frame(
  PC1=pc$PC1,
  PC2=pc$PC2,
  group=rma.data.of.interest$Status
)
ggplot(df,aes(PC1,PC2,color=group)) + geom_point() + theme_bw() + scale_color_manual("Status",values=colour_status)
```

Plot first two PCs coloured by scan date.
```{r eval=T}
df <- data.frame(
  PC1=pc$PC1,
  PC2=pc$PC2,
  group=rma.data.of.interest$ScanDate_Group
)
ggplot(df,aes(PC1,PC2,color=group)) + geom_point() + theme_bw() + scale_color_manual("Scan Date",values=colour_scandate)
```

Plot first two PCs coloured by donors if less than 10 donors were treated
```{r eval=T}
rma.data.of.interest$Subject <- as.factor(rma.data.of.interest$Subject)
if (comptype=="Treatment"&nlevels(rma.data.of.interest$Subject)<=10) {
  df <- data.frame(
    PC1=pc$PC1,
    PC2=pc$PC2,
    group=rma.data.of.interest$Subject
  )
  colour_subject <- colours[1:nlevels(rma.data.of.interest$Subject)]
  ggplot(df,aes(PC1,PC2,color=group)) + geom_point() + theme_bw() + scale_color_manual("Donor",values=colour_subject)
}
```

```{r eval=T}
sessionInfo()
```
